### PR TITLE
Update search location drawing tools for 3D map

### PR DESCRIPTION
**Pull request by jrnorth**
* Search location shapes can now be edited on the 3D map.
* A new drawing toolbar is now displayed at the top of the map. It's the same toolbar used on the 2D map.

### Testing Instructions
* Test both the basic and advanced search menus.
* Test drawing and editing all 4 shapes (polygon, bounding box, line, point-radius).
  * Drawing: draw a completely new shape.
  * Editing: drag the points/boundaries of an existing shape, and add and remove points.
    * Line: add new points by clicking and dragging the 'plus' icons, move any points on the line, and right click points to remove them.
    * Point-radius: change the radius by moving the circle's edge.
    * Bounding box: move any of the corners.
    * Polygon: add new points by clicking and dragging the 'plus' icons, move any points on the line, and right click points to remove them.
* Test entering shape coordinates manually, then editing the resulting shape.
* Test drawing and editing shapes that cross the antimeridian.
* Test cancelling the drawing via the cancel button on the toolbar and the draw button on the search menu that is toggled to a cancel button when a drawing is in progress.
* Perform some basic testing on the 2D map to ensure drawing was not broken.

### CODE REVIEW

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/lib/cesium-drawhelper/DrawHelper.ts [hunk 1]
```diff
@@ -556,22 +595,90 @@ const DrawHelper = (function () {
...
     }
     return _
   })()
+  const vertexSvg = `
+  <svg width="10" height="10" version="1.1" xmlns="http://www.w3.org/2000/svg">
+    <circle cx="5" cy="5" r="5" fill="blue"/>
+  </svg>
+  `
+  const dragHalfSvg = `
+  <svg width="8" height="8" version="1.1" xmlns="http://www.w3.org/2000/svg">
```
Comment (jrnorth) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/lib/cesium-drawhelper/DrawHelper.ts]: This SVG is a blue square with a white plus sign in the center.

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/lib/cesium-drawhelper/DrawHelper.ts [hunk 2]
```diff
@@ -556,22 +595,90 @@ const DrawHelper = (function () {
...
+        height: this.height,
+        width: this.width < 1 ? 1 : this.width,
+        vertexFormat: Cesium.EllipsoidSurfaceAppearance.VERTEX_FORMAT,
+        ellipsoid: this.ellipsoid,
+        followSurface: false,
       })
     }
     return _
   })()
+  const vertexSvg = `
```
Comment (jrnorth) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/lib/cesium-drawhelper/DrawHelper.ts]: @andrewkfiedler I want these SVGs in separate files, but I wasn't sure how to do that with the way I'm loading them in below.

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/line-display.tsx [hunk 3]
```diff
@@ -193,38 +264,125 @@ const useCameraMagnitude = ({
...
+          }
+          setNewLine(drawnLine)
+          onDraw(drawnLine)
+        },
+        material: Cesium.Material.fromType('Color', {
+          color: Cesium.Color.BLUE,
+        }),
+      })
+    }
+    return () => {}
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/line-display.tsx]: ❓ Is this empty return needed?

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/bbox-display.tsx [hunk 4]
```diff
@@ -196,12 +266,58 @@ const useListenToModel = ({ model, map }: { model: any; map: any }) => {
...
+          const bbox = rectangleToBbox(extent)
+          setNewBbox(bbox)
+          onDraw(bbox)
+        },
+        material: Cesium.Material.fromType('Color', {
+          color: Cesium.Color.fromAlpha(Cesium.Color.BLUE, 0.2),
+        }),
+      })
+    }
+    return () => {}
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/bbox-display.tsx]: ❓  Is this empty return needed?

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/circle-display.tsx [hunk 5]
```diff
@@ -181,18 +267,62 @@ const useListenToModel = ({ model, map }: { model: any; map: any }) => {
...
+          )
+          setNewCircle(circle)
+          onDraw(circle)
+        },
+        material: Cesium.Material.fromType('Color', {
+          color: Cesium.Color.fromAlpha(Cesium.Color.BLUE, 0.2),
+        }),
+      })
+    }
+    return () => {}
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/circle-display.tsx]: ❓ Is this empty return needed?

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx [hunk 6]
```diff
@@ -51,9 +77,89 @@ export const CesiumDrawings = ({
     map,
   })
 
+  const [drawingModel] =
+    drawingModels.length > 0 ? drawingModels.slice(-1) : [undefined]
+
+  const [isDrawing, setIsDrawing] = useState(false)
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx:83]: ❓ Should this be:``` const [isDrawing, setIsDrawing] = useState<boolean>(false)```

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx [hunk 7]
```diff
@@ -51,9 +77,89 @@ export const CesiumDrawings = ({
...
+      `search:${getDrawModeFromShape(drawingShape)}-end`,
+      drawingModel
+    )
+    Drawing.turnOffDrawing()
+    drawingModel.set({ ...drawingLocation, drawing: false })
+    setIsDrawing(false)
+    drawingLocation = null
+  }
+
+  console.log('filterModels', filterModels)
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx]: ❓ Are these logs just for dev, or do you really need them?

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/line-display.tsx [hunk 8]
```diff
@@ -193,38 +264,125 @@ const useCameraMagnitude = ({
...
+          })
+        }
       }
     }
-  }, [model, map])
+  }, [model, map, newLine])
   useListenTo(model, 'change:line change:lineWidth change:lineUnits', callback)
   React.useEffect(() => {
     if (map && needsRedraw({ map, drawnMagnitude })) {
+      console.log(
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/line-display.tsx]: ❓ Is this console.log() needed?

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/line-display.tsx [hunk 9]
```diff
@@ -193,38 +264,125 @@ const useCameraMagnitude = ({
...
+        '[cameraMagnitude, drawnMagnitude, callback, map] useEffect()'
+      )
       callback()
     }
   }, [cameraMagnitude, drawnMagnitude, callback, map])
   React.useEffect(() => {
+    console.log('[callback] useEffect()')
```
Comment (figliold) [inline ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/line-display.tsx]: ❓ Is this console.log() needed?

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/bbox-display.tsx (hunk 4 above)
Comment (jrnorth) [reply ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/bbox-display.tsx]: No. If our effect doesn't need any cleanup, then we don't have to return a function. I'll remove these.

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx (hunk 6 above)
Comment (jrnorth) [reply ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx:83]: We can omit the explicit type here, since TypeScript correctly infers that `isDrawing` is a `boolean` from the initial state value.

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx (hunk 7 above)
Comment (jrnorth) [reply ui-frontend/packages/catalog-ui-search/src/main/webapp/component/visualization/maps/cesium/drawing-and-display.tsx]: Do not need these, I just forgot to remove them 😄

**Comment by andrewkfiedler** [comment]
Hero went well (although my map layer wouldn't load for some reason).

Well done, this is awesome!

There was a bit of weirdness that happens if you click draw and then edit the shapes manually (it used to work but was weird honestly that we allowed it), but I wonder if we should just disable the inputs when drawing is enabled since that's a pretty weird edge-case to maintain.

I also couldn't test the removal of points on my macbook, as the ctrl (command) + click to emulate right click wasn't triggering the removal.

File: ui-frontend/packages/catalog-ui-search/src/main/webapp/lib/cesium-drawhelper/DrawHelper.ts (hunk 2 above)
Comment (andrewkfiedler) [reply ui-frontend/packages/catalog-ui-search/src/main/webapp/lib/cesium-drawhelper/DrawHelper.ts]: I pushed a commit for this https://github.com/codice/ddf-ui/pull/739/commits/e6f89b187fe75eb70b818ed7671bcda4dc24fda3

Basically you can use the raw-loader, and we can specify that it is to be used inline rather than in the webpack config (since we may not always want to import these as plain strings).

**Review by andrewkfiedler — APPROVED** [review]

**Review by alexabird — APPROVED** [review]

**Comment by alexabird** [comment]
build now
